### DISCUSSION TITLE
Launcher commands draw & view do not display text entities

### DISCUSSION THREAD

**Discussion by rdejtiar** [mozman/ezdxf#883] [Issues]
Checking a new functionality on the next release of Ezdxf I found that text is not shown on both matplotlib an pyqt5 backends .
Small rectangles shown are the text containers.
Not complaints about missing fonts .
On Pyqt5 the element is shown when selected.

![Captura de pantalla de 2023-05-14 17-42-11](https://github.com/mozman/ezdxf/assets/57047709/ac8b9110-1e37-4a66-b150-8a9957a1b2be)

**Comment by mozman** [comment]
I cannot help without the DXF file!

**Comment by mozman** [comment]
I added an option to print the available fonts that `ezdxf` found on your system: `ezdxf -f` and `ezdxf -fv`

rebuild the font cache if you added fonts:

```Python
from ezdxf.fonts import fonts

fonts.build_system_font_cache()
```

More information about fonts can be found in the docs: https://ezdxf.mozman.at/docs/tools/fonts.html

EDIT: option -f automatically rebuilds then system font cache now

**Reply by mbway** [reply]
If ezdxf does not find a font used by a DXF file does it fallback to another font or just draw nothing?

**Reply by mozman** [reply]
The following standard TTF fonts are searched (if the requested font is not available):

- Arial.ttf (should exist on Windows and macOS)
- DejaVuSansCondensed.ttf (close match to Arial.ttf)
- DejaVuSans.ttf (should exist on Linux)
- LiberationSans-Regular.ttf
- OpenSans-Regular.ttf

If none of them was found, the `MonospaceFont` measurement font is used as last resort and that font renders just empty rectangles.

All these fonts except for Arial.ttf are free available and included in the `ezdxf` repository. All tests just use these fonts from the repository, which was hard work because I didn't fully understand how `pytest` initiates tests suits. :smile:

**Comment by mozman** [comment]
@rdejtiar the distribution looks like Linux Mint, which I also use and `ezdxf` shows text as expected:

![image](https://github.com/mozman/ezdxf/assets/4230966/bb11072a-8d6d-486c-90ab-1023054dba6b)

![image](https://github.com/mozman/ezdxf/assets/4230966/ba06a6c3-d38a-46bd-8289-66dc7c029a81)

![image](https://github.com/mozman/ezdxf/assets/4230966/c7be1fb7-5c92-437d-8376-6bfb958391a3)

**Comment by rdejtiar** [comment]
Yes, I' m using Linux Mint

ezdxf  draw and  ezdxf view run showing text correctly after running ezdxf
-f (not after just installing ezdxf ver 1.1, like in ver 1.03 as far as I
remember)

I note this here just in case it is useful for somebody .

Thanks for everything .  I  appreciate your  effort  developing ezdxf for
years

**Reply by mozman** [reply]
That's strange, but reality is sometimes more complex than you think. The font cache is created automatically. In this case, an empty cache existed to prevent a cache rebuild, and forcing a rebuild with `ezdxf -f` solved the problem - at least that's the only way I can explain it.

**Comment by changkeke** [comment]
I find my dwg with Chinese char in mtext, and the mtext's content has no inner font style desc like '{\fSimsun...}', this can display correctly in autocad, but display rectangle path by mtext explode, I see the code finding used the mtext's global font style named 'Standard' of 'arial.ttf', the arial.ttf does't support Chinese char:
![image](https://github.com/mozman/ezdxf/assets/33918095/f493baf2-db0d-4fb3-9505-7e094a810c2c)
 then, I find it's hard way to modify the 'mtext's control token object code', so, I edit in 'ezdxf\tools\text.py',  the class 'MTextToken', it's construct method, I don't have any other good idea, could you give me some suggestion? code there are:
```
def is_contain_chinese(string):
    pattern = re.compile(r'[\u4e00-\u9fa5]')
    match = pattern.search(string)
    return match is not None


class MTextToken:
    __slots__ = ("type", "ctx", "data")

    def __init__(self, t: TokenType, ctx: MTextContext, data=None):
        self.type: TokenType = t
        # Handles Chinese without font control in multiple lines of text
        if ctx and data is not None:
            if is_contain_chinese(data) and ctx.font_face.family == 'Arial':
                from ezdxf.fonts.fonts import get_font_face
                ctx.font_face = get_font_face('simsun.ttc')
        self.ctx: MTextContext = ctx
        self.data = data
```

**Reply by mozman** [reply]
1. You have to check what the DXF file contains, is it a DWG to DXF conversion error?
2. Investigate why the font `SimSun.ttc` was not found: is `SimSun.ttc` available system wide? Is it included in the system font cache of `ezdxf` ?
3. Provide simple test data to check the issue.

**Reply by changkeke** [reply]
1. The screenshot above is from autocad, it's indeed missing the cahr like '{\fSimsun....}'. Here's the funny thing: when I double-click on it to try to edit it, and then finish editing, I find that its contents contain the cahr like '{\fSimsun....}'.
2. The SimSun.ttc is available in system wide, it is included in the font cache file of ezdxf.
3. The file is:
[test-mtext.zip](https://github.com/mozman/ezdxf/files/11540323/test-mtext.zip)

**Reply by mozman** [reply]
Don't send me DWG files, `ezdxf` can  read DWG via the ODAFC add-on but I will not start debugging with DWG files.

The assigned text style is `Standard` and references the `Arial` font, there are no inline codes.

![image](https://github.com/mozman/ezdxf/assets/4230966/eedd9bf6-526e-4583-87ba-cd7f17ba9984)

If you assign the `SimSun` font, `ezdxf` can display the content:
![image](https://github.com/mozman/ezdxf/assets/4230966/18cdcf3b-fa02-42c6-97c7-9b3929f1b8e4)

I don't know why BricsCAD and TrueView can display chinese characters for the `Arial` font. I assume they switch on the fly to a different font, because `Arial` does not contain chinese characters - but I don't care, just assign a proper font. There will never be an internal mapping on the fly for chinese, hebrew, arabic or other extended fonts in `ezdxf`.

**Reply by changkeke** [reply]
OK, thanks for your help! The mtext itself doesn't has valid inline code, I will fix it.
